components/renku_data_services/project/db.py

```diff
@@ -237,39 +156,42 @@ async def insert_project(
         await session.refresh(project_orm)
 
         project_dump = project_orm.dump()
-        public_project = project_dump.visibility == Visibility.public
         if project_dump.id is None:
             raise errors.BaseError(detail="The created project does not have an ID but it should.")
-        await self.project_authz.create_project(
-            requested_by=user, project_id=project_dump.id, public_project=public_project
-        )
 
         return project_dump
 
     @with_db_transaction
-    @dispatch_message(create_project_update_message)
+    @Authz.project_change(ProjectAuthzOperation.change_visibility)
+    @dispatch_message(avro_schema_v1.ProjectUpdated)
     async def update_project(
         self, session: AsyncSession, user: base_models.APIUser, project_id: str, etag: str | None = None, **payload
-    ) -> models.Project:
+    ) -> models.ProjectUpdate:
         """Update a project entry."""
-        authorized = await self.project_authz.has_permission(user=user, project_id=project_id, scope=Scope.WRITE)
-        if not authorized:
-            raise errors.MissingResourceError(
-                message=f"Project with id '{project_id}' does not exist or you do not have access to it."
-            )
-
         result = await session.scalars(select(schemas.ProjectORM).where(schemas.ProjectORM.id == project_id))
         project = result.one_or_none()
-
         if project is None:
             raise errors.MissingResourceError(message=f"The project with id '{project_id}' cannot be found")
+        old_project = project.dump()
+
+        required_scope = Scope.WRITE
+        if "visibility" in payload and (
+            (isinstance(payload["visibility"], Enum) and payload["visibility"].value != old_project.visibility.value)
+            or (
+                isinstance(payload["visibility"], str) and payload["visibility"].lower() != old_project.visibility.value
+            )
+        ):
+            required_scope = Scope.DELETE
+        authorized = self.authz.has_permission(user, ResourceType.project, project_id, required_scope)
```

Comment:
@Panaetius I am trying this but it is a bit complicated. Basically if I split this off I have to have another method where I repeat some costly operations. For example if I change the visibility and the keywords then both the `change_visibility` and `update_project` method would have to check that they are authorized to do the operation in the authorization database. Potentially I would also have to make repeated calls to postgres to to check and write things.

In addition to all of this the authz project change decorator is supposed to finalize or undo the postgres session. And in this case that is fine. A bit weird but doable.

But in the future this same `update_project` method could also potentially change the namespace of the project in addition to changing the visibility. In that case I would have to somehow have another authz decorator that either does an incremental write to the authz DB and not touch the session or the original authz decorator would have to do something to account for potentially being followed by other authz decorators that will then finalize the postgres session.

So I think it is best that I use `@Authz.project_change(ProjectAuthzOperation.change)` as the decorator rather than ``@Authz.project_change(ProjectAuthzOperation.change_visibility)`. With the PR for groups/namespaces the decorator would check if either the namespace or the visibility changed and do the right thing.